src/Group.cpp

```diff
@@ -0,0 +1,261 @@
+#include "global.h"
+#include "SongManager.h"
+#include "RageFile.h"
+#include "RageUtil.h"
+#include "RageFileManager.h"
+#include "RageLog.h"
+#include "Song.h"
+#include "SongCacheIndex.h"
+#include "SongUtil.h"
+#include "TitleSubstitution.h"
+#include "Group.h"
+
+#include <cstddef>
+#include <tuple>
+#include <vector>
+
+
+/** @brief The file that contains the group information.
+ * We name this Pack.ini over Group.ini to avoid conflict
+ * with the Waiei Group.ini-lua project still actively
+ * being developed.
+*/
+const RString INI_FILE = "Pack.ini";
+const int INI_VERSION = 1;
+
+Group::Group() {
+    m_sDisplayTitle = "";
+    m_sSortTitle = "";
+    m_sPath = "";
+    m_sGroupName = "";
+    m_sTranslitTitle = "";
+    m_sSeries = "";
+    m_fSyncOffset = 0;
+    m_bHasPackIni = false;
+    m_iYearReleased = 0;
+    m_sBannerPath = "";
+}
+
+Group::~Group() {
+    SONGMAN->GetGroupGroupMap().clear();
+}
+
+Group::Group(const RString& sDir, const RString& sGroupDirName) {
+    RString sPackIniPath = sDir + sGroupDirName + "/" + INI_FILE;
+    RString sDisplayTitle = sGroupDirName;
+    RString sSortTitle = sGroupDirName;
+    RString sTranslitTitle = "";
+    RString sSeries = "";
+    RString sBannerPath = "";
+    float fOffset = PREFSMAN->m_fMachineSyncBias;
+
+    if (FILEMAN->DoesFileExist(sPackIniPath)) {
+        IniFile ini;
+        ini.ReadFile(sPackIniPath);
+        ini.GetValue("Group", "DisplayTitle", sDisplayTitle);
+        Trim(sDisplayTitle);
+        if (sDisplayTitle.empty()) {
+            sDisplayTitle = Basename(sGroupDirName);
+        }
+        ini.GetValue("Group", "Banner", sBannerPath);
+        ini.GetValue("Group", "SortTitle", sSortTitle);
+        Trim(sSortTitle);
+        if (sSortTitle.empty()) {
+            sSortTitle = sDisplayTitle;
+        }
+        ini.GetValue("Group", "TranslitTitle", sTranslitTitle);
+        Trim(sTranslitTitle);
+        if (sTranslitTitle.empty()) {
+            sTranslitTitle = sDisplayTitle;
+        }
+        ini.GetValue("Group", "Series", sSeries);
+        RString sValue = "";
+        
+        ini.GetValue("Group", "SyncOffset", sValue);
+        Trim(sValue);
+        if (!sValue.empty()) {
+            if (sValue.CompareNoCase("null") == 0) {
+                fOffset = 0.0f;
+            } else if (sValue.CompareNoCase("itg") == 0) {
+                fOffset = -0.009f;
+            } else {
+                fOffset = StringToFloat(sValue);
+            }
+        }
+        ini.GetValue("Group", "Year", m_iYearReleased);
+        
+        RString sVersion = "";
+        ini.GetValue("Group", "Version", sVersion);
+        Trim(sVersion);
+        if (!sVersion.empty()) {
+            m_iVersion = StringToInt(sVersion);
+        } else {
+            // If the version is not set, set it to the current version
+            // and write it to the file.
```

Comment:
I think this is kind of bad. If someone has release a pack without a version number, then the version it gets set to is completely depending on like the itgmania version.

In this case we can just try our best, but we shouldn't overwrite the file. Just log a warning here saying that it may have been loaded incorrectly. Alternatively, we can check this first and just not parse the file at all if it's not set (also log a warning).